cmd/celestia/rpc.go

```diff
@@ -0,0 +1,284 @@
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/spf13/cobra"
+
+	"github.com/celestiaorg/celestia-node/api/rpc/client"
+	"github.com/celestiaorg/celestia-node/state"
+)
+
+const (
+	authEnvKey = "CELESTIA_NODE_AUTH_TOKEN"
+)
+
+var requestURL string
+var authTokenFlag string
+
+type jsonRPCRequest struct {
+	ID      int64         `json:"id"`
+	JSONRPC string        `json:"jsonrpc"`
+	Method  string        `json:"method"`
+	Params  []interface{} `json:"params"`
+}
+
+func init() {
+	rpcCmd.PersistentFlags().StringVar(
+		&requestURL,
+		"url",
+		"http://localhost:26658",
+		"Request URL",
+	)
+	rpcCmd.PersistentFlags().StringVar(
+		&authTokenFlag,
+		"auth",
+		"",
+		"Authorization token (if not provided, the "+authEnvKey+" environment variable will be used)",
+	)
+	rootCmd.AddCommand(rpcCmd)
+}
+
+var rpcCmd = &cobra.Command{
+	Use:   "rpc [namespace] [method] [params...]",
+	Short: "Send JSON-RPC request",
+	Args:  cobra.MinimumNArgs(2),
+	ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		modules := client.Modules
+		if len(args) == 0 {
+			// get keys from modules (map[string]interface{})
+			var keys []string
+			for k := range modules {
+				keys = append(keys, k)
+			}
+			return keys, cobra.ShellCompDirectiveNoFileComp
+		} else if len(args) == 1 {
+			// get methods from module
+			module := modules[args[0]]
+			methods := reflect.VisibleFields(reflect.TypeOf(module).Elem())
+			var methodNames []string
+			for _, m := range methods {
+				methodNames = append(methodNames, m.Name)
+			}
+			return methodNames, cobra.ShellCompDirectiveNoFileComp
+		}
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		namespace := args[0]
+		method := args[1]
+		params := parseParams(method, args[2:])
+
+		sendJSONRPCRequest(namespace, method, params)
+	},
+}
+
+func parseParams(method string, params []string) []interface{} {
```

Comment:
The fact that we still need to switch over most of the commands kinda justifies having a separate command for each. Although, I am ok with current approach.

Reply:
The possible issue I see is code scalability. Like whether the API changes or we add options to the API, changing this parse func would be painful

Reply:
I see. The scalability point is still there